src/server.c

```diff
@@ -4693,17 +4693,27 @@ void resetErrorTableStats(void) {
 void redisOpArrayInit(redisOpArray *oa) {
     oa->ops = NULL;
     oa->numops = 0;
+    oa->capacity = 0;
 }
 
 int redisOpArrayAppend(redisOpArray *oa, int dbid, robj **argv, int argc, int target) {
     redisOp *op;
+    int prev_capacity = oa->capacity;
 
-    oa->ops = zrealloc(oa->ops,sizeof(redisOp)*(oa->numops+1));
-    op = oa->ops+oa->numops;
+    if (oa->numops == 0) {
+        oa->capacity = 16;
+    } else if (oa->numops >= oa->capacity) {
+        oa->capacity *= 2;
+    }
+
+    if (prev_capacity != oa->capacity)
+        oa->ops = zrealloc(oa->ops,sizeof(redisOp *)*oa->capacity);
+    op = zmalloc(sizeof(*op));
```

Comment:
Why did you turn it from an op array to an op pointer array? 
So that the realloc is smaller (less memcpy)?

Reply:
yes

Reply:
i'm not sure it's a win (doing more calls to malloc). unless the array is expected to be huge, which i doubt.
i.e. you could have added the greedy growth mechanism, and keep it a flat array.

Reply:
when running `LATENCY of expire events are correctly collected` with address-sanitizer it made a huge difference (this test uses a lua to add 1M elements to a set. which means the also_propagate array reaches 1M*sizeof(redisOp), which is 64MB, instead of just 8MB when we store pointers.

maybe the hole scenario is a bit extreme (address sanitizer + creating a 1M member set in a lua)

your call

Reply:
This huge difference is between the version with and without the extra pointer? or from before you added the greedy allocation? 
I don't understand why address sanitizer will be slow to realloc more memory. If anything the extra indirection has an extra cost.

Reply:
i re-checked and you're right, the greedy allocation was the game changer... I'll keep only the greedy allocation